x/mongo/driver/topology/server_test.go

```diff
@@ -269,6 +270,82 @@ func TestServerConnectionTimeout(t *testing.T) {
 	}
 }
 
+func TestServerConnectionCancellation(t *testing.T) {
+	// Start a goroutine that pulls events from the events channel and inserts them into the
+	// events slice. Use a sync.WaitGroup to allow the test code to block until the events
+	// channel loop exits, guaranteeing that all events were copied from the channel.
+	events := make([]*event.PoolEvent, 0)
+	eventsCh := make(chan *event.PoolEvent)
+	var eventsWg sync.WaitGroup
+	eventsWg.Add(1)
+	go func() {
+		defer eventsWg.Done()
+		for evt := range eventsCh {
+			events = append(events, evt)
+		}
+	}()
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer func() {
+		_ = l.Close()
+	}()
+
+	// Create a context with cancellation that can be cancelled before the dial completes.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	server, err := NewServer(
+		address.Address(l.Addr().String()),
+		primitive.NewObjectID(),
+		// Create a connection pool event monitor that sends all events to an events channel
+		// so we can assert on the connection pool events later.
+		WithConnectionPoolMonitor(func(_ *event.PoolMonitor) *event.PoolMonitor {
+			return &event.PoolMonitor{
```

Comment:
Optional: IIUC this could use the new `testPoolMonitor`.

Reply:
Yes it could, although it would require moving `testPoolMonitor` into a test util package like `internal/testutil` (or even an external test utilities package if that makes sense). I'll leave a comment to indicate that change should be considered as part of GODRIVER-2068.